### ISSUE TITLE
Add std::coroutine_max_size_v<Callable, T1, T2, ...> magic template variable

### ISSUE THREAD

**Issue by GorNishanov** [modocache/clang#9]
One of the strong feedback for Coroutine TS was the desire to have an ability to synthesize coroutines that are guarantee not to allocate any memory even in debug mode. This template variable implemented by the compiler extends the existing Coroutine TS by offering an ability to create on stack coroutines.

[Related issue to be handled by Gor, middle end assert guaranteeing a compilation error if the actual coroutine size ended up bigger than the front end estimate]

**Comment by lewissbaker** [comment]
Is the idea that this would be a constexpr constant that we could use to allocate a `std::array<char, std::coroutine_max_size_v<Callable, T1, T2, ...>>` or a non-constexpr static variable only queryable at runtime that we could use to preallocate dynamically enough storage?

How would this cope with recursive functions? Would this have an undefined result?
eg.
```
task<void> f(std::allocator_arg_t, stack_allocator allocator, int n)
{
  if (n == 0) co_return;
  static constexpr auto coroSize =
    std::coroutine_max_size_v<f, std::allocator_arg_t, stack_allocator, int>;
  char buffer[coroSize];
  co_await f(std::allocator_arg, stack_allocator{buffer, coroSize}, n - 1);
}
```

**Comment by GorNishanov** [comment]
Yes. It is a compile time constant. Used in combination with a contract on operator new of the promise to do a middle end time assert that the size fits. I am thinking of using the presence of that assert should disable Halo optimization for the function, hence, recursion should not be a problem.

In general, any attempt to manifest sizeof of coroutine state, must disable Halo, or else, it will be a lie.

**Comment by GorNishanov** [comment]
Never mind my previous answer. I was thinking purely in terms of embedding coroutine in the return type as in: https://wandbox.org/permlink/nGNjY6cELAkvIOLC

For the case you described, this should be ill-formed. Not sure yet how to specify it.

**Comment by GorNishanov** [comment]
I don't think this is so much about recursion, but, using sizeof of itself while calculating its own size

**Comment by GorNishanov** [comment]
When we are asking "std::coroutine_max_size_v<f, std::allocator_arg_t, stack_allocator, int>;" f does not yet have a definition, hence, it is ill-formed. It has only a declaration.

**Comment by lewissbaker** [comment]
See https://github.com/modocache/clang/issues/10 for a different approach that could use the type-system.

Using the type-system but deferring the layout seems like a preferable direction, if that approach is actually viable, since it still allows the compiler to fully optimise the coroutine frame size without needing to offer up an approximate/conservative frame size to the front-end.

**Comment by GorNishanov** [comment]
Preferable direction is not to expose any estimate of the coroutine frame size as compile time constant. (That is the status quo). However, we need to be responsive to concerns expressed and do due diligence to demonstrate possible solutions, even if they won't be taken in the end.

**Comment by modocache** [comment]
Quick progress update: as far as I understand this, the semantics would be similar to the `sizeof` operator in C++, in that it would return the size of some type (or set of types) at compile time:

```cpp
struct x { int a; float b; double c; };

int main() {
  return sizeof(x); // 16
}
```

`sizeof` is represented in the Clang AST with `UnaryExprOrTypeTraitExpr`. During LLVM IR codegen for the above program, `HandleSizeof` calls `ASTContext::getTypeInfoInChars` with the type argument `x`.

To implement `std::coroutine_max_size_v<...>`, I'll try calling `ASTContext::getTypeInfoInChars` with the potential members of the coroutine frame type. I believe these become known during the coro-early LLVM pass? I'm hoping to have this figured out tomorrow.

**Comment by GorNishanov** [comment]
Short answer:
std::coroutine_max_size_v<...> should be implemented fully during Sema phase of the frontend.
1. To do that, we need to find the definition of callable when instantiated/called with the arguments types provided. 
2. FunctionDefinition should have `CoroutineBodyStmt` as a member. CoroutineBodyStmt should have all possible pieces known to the frontend that will go into the coroutine body. Summing up all of the sizes of possible locals/temporaries with automatic storage duration should give a rough estimate of the upper bound of the size of the coroutine frame.
3. We also know what extra stuff is added to Coroutine Frame by llvm passes, namely, couple of pointers + coroutine state index + some padding.
4. Probably when computing size in step 2, we can exclude any local/temporaries from blocks that do not have await/yield expressions.
5. Step 4 is at the moment is not reliable, in cases when address of a local escapes the coroutine, as CoroFrame analysis in CoroSplit.cpp blindly assumes that anything that has an address taken and that address escapes the coroutine should live in Coroutine Frame (since potentially it can be accessed across suspend point. To make step 4 reliable, we need to pay attention to lifetime markers)

Long answer (ended up being shorter somehow :-) ):
Due to "coroutine uncertainty principle", coroutine can be either fast or you can know its precise size :-).
Namely, sizeof is something known to frontend, actual coroutine size is known in the middle end (or even later, in case of Link Time Optimization. Unless we intend to run compiler in a loop and feed information from the middle end back to the frontend until the coroutine sizes estimates will be within some tolerable margin, we cannot now the exact size of the coroutine state. This process potentially can never converge as whenever size of the coroutine frame changes, something can get inlined and make coroutine size bigger and thus have to repeat the cycle again.

That is related to the fact that to make coroutines efficient we need to defer splitting the coroutine until the middle end, so that all traditional function optimizations can apply before doing the split. Knowing the size is not a problem if Coroutine Transformation is fully performed in the frontend, but, those coroutines unlikely to have as efficient codegen as the coroutines which are split in the middle end.

**Comment by modocache** [comment]
Got it, thanks @GorNishanov! That sounds like a great approach, and I'm fairly familiar with libclangSema and the `CoroutineBodyStmt`, so it shouldn't be too difficult to iterate over the statements in the `CoroutineBodyStmt::Body` and sum their sizes, the sizes of the coroutine's parameters, and the stuff I know will be added by the LLVM passes.

I'll try to have something working, excluding step 4, by tonight or tomorrow.

**Comment by GorNishanov** [comment]
Cool! I am working on adding the compile time check for coroutine size to llvm. Probably tonight or tomorrow, I will send a pull request.

Since contracts are not yet available in clang, we can use __attribute__((diagnose_if(sz >= N, "", "error"))); instead. SemaCoroutine/CGCoroutine can map that attribute (if present on operator new of the coroutine promise) to emit the llvm intrinsic that will do the check in CoroSplit once the coroutine size is known.

I'll file an issue to hookup to diagnose_if, once llvm intrinsic is available.

**Comment by modocache** [comment]
For now I have something like this:

```diff
diff --git a/lib/Sema/SemaCoroutine.cpp b/lib/Sema/SemaCoroutine.cpp
index 9c9d266d3d..abebc41336 100644
--- a/lib/Sema/SemaCoroutine.cpp
+++ b/lib/Sema/SemaCoroutine.cpp
@@ -18,12 +18,16 @@
 #include "clang/AST/ASTLambda.h"
 #include "clang/AST/Decl.h"
 #include "clang/AST/ExprCXX.h"
+#include "clang/AST/RecursiveASTVisitor.h"
 #include "clang/AST/StmtCXX.h"
 #include "clang/Lex/Preprocessor.h"
 #include "clang/Sema/Initialization.h"
 #include "clang/Sema/Overload.h"
 #include "clang/Sema/ScopeInfo.h"
 #include "clang/Sema/SemaInternal.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "coro-max-size"
 
 using namespace clang;
 using namespace sema;
@@ -930,6 +934,45 @@ static FunctionDecl *findDeleteForPromise(Sema &S, SourceLocation Loc,
   return OperatorDelete;
 }
 
+static uint64_t getCoroutineMaxSizeV(Sema &S, FunctionDecl *FD, Stmt *Body) {
+  assert(isa<CoroutineBodyStmt>(Body) && "Not a coroutine body");
+
+  // Parameters to the coroutine spill across the initial suspend point
+  // and so may need to be stored in the coroutine frame.
+  uint64_t ParamSize = 0;
+  for (auto *PD : FD->parameters()) {
+    CharUnits Units = S.getASTContext().getTypeSizeInChars(PD->getType());
+    ParamSize += Units.getQuantity();
+  }
+
+  // Traverses variable declarations and sums their size.
+  class VarSizeVisitor : public RecursiveASTVisitor<VarSizeVisitor> {
+    const ASTContext &Context;
+    uint64_t Size = 0;
+
+  public:
+    VarSizeVisitor(ASTContext &Context) : Context(Context) {}
+
+    uint64_t getSize() const { return Size; }
+
+    bool VisitVarDecl(VarDecl *VD) {
+      CharUnits Units = Context.getTypeSizeInChars(VD->getType());
+      Size += Units.getQuantity();
+      return true;
+    }
+  };
+
+  // Local and temporary variables may spill across suspend points
+  // and so may need to be stored in the coroutine frame.
+  VarSizeVisitor V(S.getASTContext());
+  V.TraverseStmt(Body);
+  uint64_t VarSize = V.getSize();
+
+  // TODO: Add size of pointers, coroutine resume index, and other data we know
+  //       will be added by LLVM.
+
+  return ParamSize + VarSize;
+}
 
 void Sema::CheckCompletedCoroutineBody(FunctionDecl *FD, Stmt *&Body) {
   FunctionScopeInfo *Fn = getCurFunction();
@@ -964,6 +1007,9 @@ void Sema::CheckCompletedCoroutineBody(FunctionDecl *FD, Stmt *&Body) {
 
   // Build body for the coroutine wrapper statement.
   Body = CoroutineBodyStmt::Create(Context, Builder);
+
+  LLVM_DEBUG(llvm::dbgs() << "Max size for '" << FD->getNameAsString() << "': "
+                          << getCoroutineMaxSizeV(*this, FD, Body) << "\n");
 }
 
 CoroutineStmtBuilder::CoroutineStmtBuilder(Sema &S, FunctionDecl &FD,
```

Which for the following program:

```cpp
struct awaitable {
  bool await_ready() { return true; }
  void await_suspend(void *) {}
  void await_resume() {}
};

namespace std {
namespace experimental {
template<typename T> struct coroutine_handle {
  static void *from_address(void *address) { return nullptr; }
};

template<typename ReturnType, typename... ArgTypes> struct coroutine_traits {
  struct promise_type {
    awaitable initial_suspend() { return {}; }
    awaitable final_suspend() { return {}; }
    void return_void() {}
    void get_return_object() {}
    void unhandled_exception() {}
  };
};
}
}

struct A { int a; float b; double c; };
template<typename T> struct B { T a; };

void foo(A a, int b, float c, double d, void *e) {
  int f = 42;
  A g{};
  B<int> h = {};
  float i = d + g.b + h.a;
  auto j = e;
  auto *k = e;
  co_return;
}

int main() {
  foo(A{}, 1, 2.f, 3.0, nullptr);
}
```

Returns this:

```
clang++ -fcoroutines-ts -stdlib=libc++ -c example.cpp -mllvm -debug-only=coro-max-size
Max size for 'foo': 84
```

This checks out, I think:

```
sizeof(A)      => 16 // parameter 'a'
sizeof(int)    => 4  // parameter 'b'
sizeof(float)  => 4  // parameter 'c'
sizeof(double) => 8  // parameter 'd'
sizeof(void*)  => 8  // parameter 'e'
sizeof(int)    => 4  // variable 'f'
sizeof(A)      => 16 // variable 'g'
sizeof(B<int>) => 4  // variable 'h'
sizeof(void*)  => 4  // variable 'j'
sizeof(void*)  => 4  // variable 'k'
--------------------
Total          => 84
```

I still need to:
1. Figure out exactly what to add for (3), the stuff that's added to the frame by LLVM passes.
2. Implement this as a template variable. Is there a similar "magic" template variable that's already implemented by Clang?

**Comment by GorNishanov** [comment]
Not sure about template variable, but, there are a lot of traits implemented as compiler intrinsics / builtins. Maybe that can give an idea.

There is also a code in SemaCoroutine.cpp that constructs instantiation of coroutine_traits<stuff>. That could be helpful.

Also, as a sanity check, add

````cpp
operator new(size_t sz) { 
  printf("%zd\n", sz);
  return ::operator new(sz);
}
````

to the coroutine promise to check how accurate is your estimate. I suspect that there are coroutine_handles, awaitables (initial_suspend, final_suspend) and (Gro Initialized) that might need to get added up too.

**Comment by modocache** [comment]
Thanks! Yes, I figured the `<type_traits>` header would be a good reference, but as far as I know, those template variables, and `std::experimental::coroutine_traits`, are all defined in libcxx. I guess I'll have to do the same: declare the variable `std::experimental::coroutine_max_size_v` in `<experimental/coroutine>`, and in its implementation call a Clang builtin that implements similar logic to the above -- or something. I'll figure this out today.

Good point also about the `__coro_gro` and `__promise` variables, etc., I'll make sure to add those to the sum as well.

**Comment by GorNishanov** [comment]
Or, and more, and more. clang is doing very verbose cleanup code (for any function) that includes some cleanup variable. 

Doing `-S -emit-llvm -disable-llvm-passes` and looking at all of the alloca should tell you all of the cruft that clang inserts into a function.

**Comment by lewissbaker** [comment]
Hi! This is Gor pretending to be Lewis. I think there is a straightforward way of doing the template variable. A bit of lib + a clang builtin.

You implement the variable fully in the library and to get the value you call the built-in.

__builtin_max_size(&Callable(declval&lt;T1&gt;(), declval&lt;T2&gt;(), ...) // or possibly invoke to deal with member functions (Thank you Lewis).

That should run the instantiation and you will have a nice AST node pointing at what you need.

**Comment by modocache** [comment]
Thanks! That's a good idea -- I was also thinking of having the library call a builtin that would do the actual summation, but before reading this I was working on implementing a builtin template similar to `__make_integer_seq`. I think what I was missing was the builtin's signature -- in the example you posted, for instance, what is the type signature of `__builtin_max_size`? Specifically, what are the types of its arguments?

(Also, be careful @lewissbaker, it sounds like @GorNishanov has your password! 😆)

**Comment by GorNishanov** [comment]
__builtin_max_coroutine_size(void*)

**Comment by modocache** [comment]
Just some more clarification, if you please: let's pretend the end user calls the builtin directly. To pass in the `foo` coroutine, would this work...? https://godbolt.org/z/PNHFnA

Basically I'm not proficient enough in templates to understand how to pass a coroutine `foo`, resolved using a given set of argument types, into the builtin. Could you post a godbolt link that shows an example of the builtin being called?

**Comment by GorNishanov** [comment]
Something like this: https://godbolt.org/z/3jHwGm ?

**Comment by lewissbaker** [comment]
The problem with that approach is that it only works for function objects.

If we have a normal free function with an overload set then we can't specify a 'Callable' that exactly identifies a particular overload. e.g. https://godbolt.org/z/jnpuY6

The 'Callable' would need to be type-erased to a function-pointer of the appropriate signature which would throw away the type information required to identify a specific coroutine.
eg. How do we identify through the type-system the different overloads of f and between the overload of f and g that have the same signature?
```c++
task<void> f(int);
task<void> f(bool);
task<void> g(int);
```

I think we would instead need some kind of 'decltype()' style operator where the argument expression is required to be a call expression to some function which is required to be a coroutine.
eg.
```c++
task my_type::some_method();
auto lambda = [](int) -> task { ... };

size_t x = __builtin_coroutine_max_size(f(std::declval<int>()));
size_t y = __builtin_coroutine_max_size(f(std::declval<bool>()));
size_t z = __builtin_coroutine_max_size(g(std::declval<int>()));
size_t w = __builtin_coroutine_max_size(lambda(std::declval<int>()));
size_t u = __builtin_coroutine_max_size(std::declval<my_type>().some_method());
```

**Comment by GorNishanov** [comment]
Sure. At the moment, though, the only known use case is asking for size of a coroutine lambda. 

````c++
auto f(int a, int b) {
  return make_on_stack([&](auto) -> generator<int> { co_yield a + b; });
}
````
I am looking for a way with minimal time investment to make the example above work. (Note that `make_on_stack` is a normal function using the `std::coroutine_max_size_v` under the covers.
The approach I am suggesting is do some kind of a library trick along the lines I showed above so that clang will automatically produce an AST node from which Brian can get a reference to instantiated body of the coroutine so that the size of the locals can be computed.

I have very limited clang expertise, so any suggestion I am giving reflects how I would try to implement something like given the constraints we have.

Based on the feedback to "Understanding allocation of Coroutine State" paper that will offer this + contract approach, we can re-evaluated whether we need more sophisticated solution/implementation.

For example, with more time, it could be:

````c++
co_sizeof(f(std::declval<int>()));
````

But I don't find it aesthetically pleasing. Or:

````c++
sizeof(max_coroutine_size(f(std::declval<int>()));
````
That also does not look right, since `f` might be returning void.

**Comment by modocache** [comment]
Sorry this is taking longer than I originally anticipated! My current status is that I'm reading https://godbolt.org/z/3jHwGm and trying to understand it well enough to translate it into a template builtin function in Clang.

**Comment by GorNishanov** [comment]
I am up for any in language/lib construct that will give me an estimate of the size of the callable. If, for example, co_sizeof(Callable(decltype<T1>(), decltype<T2>(), ...) is easier to implement, that would work too.

**Comment by modocache** [comment]
I feel I'm close; with this patch https://reviews.llvm.org/P8119 applied to Clang, with C++ program https://reviews.llvm.org/P8120, prints the sizes `4` and `84` for the two lambdas. I think the only thing left is to add the maximum size of the LLVM frame struct -- its `int8` resume index, the two resume function pointers, and the promise. I don't yet know how to get those sizes, but I'm close I think.

**Comment by GorNishanov** [comment]
That is a special knowledge that frontend needs to have.

sizeof(void*) * 2 + pading so that promise is properly aligned + sizeof(promise) + int64 (for index)

**Comment by modocache** [comment]
I made some more progress with this. I have some work locally that stores the constructs you describe, but I need to tweak the logic I use to detect spills -- as you described here:

> Summing up all of the sizes of possible locals/temporaries with automatic storage duration should give a rough estimate of the upper bound of the size of the coroutine frame.

Right now I only account for `VarDecl`, which misses temporaries such as, using the types from the example program above, `int x = A{}.a;`. I also miss the temporaries generated for the coroutine handles that are passed into calls to `await_resume`, as well as the awaitables implicitly generated at the `co_await` points for initial and final suspend. Still, I believe that if I sum up the sizes for all `MaterializeTemporaryExpr` , I'll have arrived at the true upper bound. I'm hoping to post a full patch soon.

**Comment by modocache** [comment]
I sent up https://reviews.llvm.org/D56243 -- sorry this took me a while!
